mkosi/__init__.py

```diff
@@ -2127,11 +2172,17 @@ def run_qemu(args: MkosiArgs, config: MkosiConfig) -> None:
             elif config.architecture == "aarch64":
                 cmdline += ["-device", "tpm-tis-device,tpmdev=tpm0"]
 
+        addr, notifications = stack.enter_context(vsock_notify_handler())
+        cmdline += ["-smbios", f"type=11,value=io.systemd.credential:vmm.notify_socket={addr}"]
+
         cmdline += config.qemu_args
         cmdline += args.cmdline
 
         run(cmdline, stdin=sys.stdin, stdout=sys.stdout, env=os.environ, log=False)
 
+    if "EXIT_STATUS" in notifications:
+        sys.exit(int(notifications["EXIT_STATUS"]))
```

Comment:
Right now `run_qemu` (and `run_ssh` and `run_serve`) fall through to the end of `run_verb`. We removed the parts that `exec`ed so that exits where handled in a single place and while this already pretty much the end, returning the exit code and propagating returning it via `run_verb` and then doing the `sys.exit` in `__main__` would be nicer, since the function can then also be called in isolation without having to catch the `SystemExit`.

Reply:
I'm not a fan of returning the exit code. I agree that calling sys.exit() is not very nice though, so I changed it to raise subprocess.CalledProcessError() instead, so it follows exactly the same codepath as when another command fails.